### PR TITLE
Use BigRational for LSeq Identifiers and Rename LSeq to List

### PR DESCRIPTION
**Pull request by davidrusu**
LSeq Identifiers as implemented faithfully to the papers have some undesirable properties, namely around their sensitivity to mis-used allocation strategies.

In practice we were seeing panics due to the exponential tree growing beyond our max height.

This PR:
1. replaces our Identifiers with the tuple `(BigRationals, Dot)`
2. Expands the interval for allowable identifiers from (0,1) to all of the rationals (this is more efficient given how BigRationals are represented internally)
3. The allocation strategy is now a deterministic algorithm, we simply choose the midpoint between identifiers we would like to insert between.

### CODE REVIEW

File: src/lseq.rs [hunk 1]
```diff
@@ -36,95 +36,103 @@
...
-    pub fn new(id: A) -> Self {
-        LSeq {
-            seq: Vec::new(),
-            gen: IdentGen::new(id),
-            clock: VClock::new(),
+impl<T, A: Ord> Default for LSeq<T, A> {
+    fn default() -> Self {
+        Self {
+            seq: Default::default(),
+            clock: Default::default(),
```
Comment (bochaco) [inline src/lseq.rs:127]: What's your opinion on not using `Default::default() but the specific type's `default()`? just curious as I thought it'd be better and more strict to use the specific type's one but...not sure...I guess it's not really a big deal in reality

File: src/lseq.rs [hunk 2]
```diff
@@ -133,71 +141,59 @@ impl<T: Clone, A: Ord + Clone> LSeq<T, A> {
     /// # Panics
     ///
     /// * If the allocation of a new index was not between `ix` and `ix - 1`.
```
Comment (bochaco) [inline src/lseq.rs:143]: Does this still panic in such case?

File: src/lseq.rs [hunk 3]
```diff
@@ -133,71 +141,59 @@ impl<T: Clone, A: Ord + Clone> LSeq<T, A> {
...
-
-        Op::Insert {
-            id: ix_ident,
-            dot: self.clock.inc(self.actor()),
-            val,
-        }
+        let id = match (prev, next) {
+            (Some(p), Some(q)) => (&p.id + &q.id) / two,
+            (Some(p), None) => &p.id + &one,
+            (None, Some(p)) => &p.id - &one,
```
Comment (bochaco) [inline src/lseq.rs:166]: Perhaps for more readability make it `(None, Some(q))`?

File: src/lseq.rs [hunk 4]
```diff
@@ -133,71 +141,59 @@ impl<T: Clone, A: Ord + Clone> LSeq<T, A> {
...
+        let index = Index { id, dot };
+        Op::Insert { index, val }
     }
 
     /// Perform a local insertion of an element at the end of the sequence.
-    pub fn append(&mut self, c: T) -> Op<T, A> {
+    pub fn append(&mut self, c: T, actor: A) -> Op<T, A> {
         let ix = self.seq.len();
-        self.insert_index(ix, c)
+        self.insert_index(ix, c, actor)
```
Comment (bochaco) [inline src/lseq.rs:177]: Why the actor is not fixed to the instance? any replica would create an instance with its own actor, and local operation shall/will always be made with that actor whilst other's actor ops will be "applied" intead?

File: src/lseq.rs (hunk 1 above)
Comment (davidrusu) [reply src/lseq.rs:127]: I get the argument for explicitness, but I feel at some point you need to trust the abstractions. The types are defined explicitly at the struct definition, here all I want to say is I want the default value of these fields, no matter what the type changes to in the future.

File: src/lseq.rs (hunk 2 above)
Comment (davidrusu) [reply src/lseq.rs]: It does not! I've gone ahead and updated all the docs

File: src/lseq.rs (hunk 3 above)
Comment (davidrusu) [reply src/lseq.rs]: Yes, good point

File: src/lseq.rs (hunk 4 above)
Comment (davidrusu) [reply src/lseq.rs]: I don't think CRDT's should remember who their owners are, my view is that Identity should be managed by the application logic. Also, this is looking ahead a bit, I'd like to have Lists embed-able in Map in the future and we need to be able to construct a Default instance of List to do this. (You can have default if the constructor depends on the actor)

**Comment by iancoleman** [comment]
This removed bitvec but it is still listed in Cargo.toml

**Comment by davidrusu** [comment]
@iancoleman thanks for catching that, removed in #108

**Pull request merged**